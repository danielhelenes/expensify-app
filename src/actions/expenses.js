//since we don't have any jsx, we don't need to import react.

import uuid from 'uuid';

 export const addExpense = ({description = '', note = '', amount = 0, createdAt = 0} = {}) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id: uuid(),
    description,
    note, // this is the same of note: note
    amount,
    createdAt
  }
})

/*
// action above is the same as the one below:
export const addExpense = (anything = {}) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id: uuid(),
    description: anything.description,
    note: anything.note, // this is the same of note: note
    amount: anything.amount,
    createdAt: anything.createdAt
  }
})*/


// REMOVE EXPENSES

export const removeExpense = ({id} = {}) => ({ //question: Why do we need the default here?
  type: 'REMOVE_EXPENSE',
  id
})

// EDIT EXPENSES

export const editExpense = (id, updates) => ({ //we don't need defaults here.
  type: 'EDIT_EXPENSE',
  id,
  updates
});
