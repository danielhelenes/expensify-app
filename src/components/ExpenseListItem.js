// export a stateless functional component
// description, amount, createdAt
// use map in expenseList.js

import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({ id, description, amount, createdAt, note }) =>  ( //props is not defined. console logs each expenses object with amount, createdat, etc... but not expenses in general  (it lists all the objects of items)
  <div>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>
      {numeral(amount / 100).format('$0,0.00')}
      -
      {moment(createdAt).format('Do MMMM, YYYY')}
    </p>
  </div>
);

export default ExpenseListItem;



// -----------------------------//////////////// -----------------------------//////////////


//Same as above:
// const ExpenseListItem = (props) =>  ( //props is not defined. console logs each expenses object with amount, createdat, etc... but not expenses in general  (it lists all the objects of items)
//   <div>
//     <Link to={`/edit/${props.id}`}>
//       <h3>{props.description}</h3>
//     </Link>
//     <p>{props.amount} - {props.createdAt}</p>
//   </div>
// );

// MY SOLUTION => passing mapStateToProps!

// const ExpenseListItem = (props) => (
//   <div>
//     <h3>{props.description}</h3>
//     <h3>{props.id}</h3>
//     <p>{props.amount} - {props.createdAt}</p>
//     <button onClick={() => props.dispatch(removeExpense({id: props.id}))}>Remove</button>
//   </div>
// );
//
// const mapStateToProps = (state) => {
//   return {
//     expenses: state.expenses
//   }
// };
//
// export default connect(mapStateToProps)(ExpenseListItem);

// -----------------------------//////////////// -----------------------------//////////////
