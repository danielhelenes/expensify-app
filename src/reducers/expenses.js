// Expenses Reducer - gets all the expenses related components

//question: why does this reducer has access to action if they are in different folders and there is no import here??!?!

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [...state, action.expense];
    case 'REMOVE_EXPENSE':
      return state.filter((expense) => {//this is the same as cur. we map through the cur object that state has. we look for the exact id inside of all the objects that has the same id of the one we are deleting, so we can delete it.
        return expense.id !== action.id;
      });

    case 'EDIT_EXPENSE':
    return state.map((expense) => { //this is the same as cur. we map through the cur object that state has. if the id of each cur object matches the id of the action object (the one we are editing), we return the cur object + the updates.
      if (expense.id === action.id) {
        return {...expense, ...action.updates}
      }else { //for all the objects that have a different id than the one we are editing, we return the same object. nothing changes.
        return expense;
      }
    });

    default:
      return state;
  }
};


///IMPORTANT: the state is simply what you add to it. that is the following object:
/*expense: {
  id: uuid(),
  description,
  note, // this is the same of note: note
  amount,
  createdAt
}*/
