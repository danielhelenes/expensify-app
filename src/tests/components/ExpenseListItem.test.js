import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

test('should render expense list item correctly', () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[1]}/>);
  expect(wrapper).toMatchSnapshot();

});











//
// const ExpenseListItem = ({ id, description, amount, createdAt, note }) =>  ( //props is not defined. console logs each expenses object with amount, createdat, etc... but not expenses in general  (it lists all the objects of items)
//   <div>
//     <Link to={`/edit/${id}`}>
//       <h3>{description}</h3>
//     </Link>
//     <p>{amount} - {createdAt} / {note && `/ ${note}`}</p>
//   </div>
// );
//
// export default ExpenseListItem;
