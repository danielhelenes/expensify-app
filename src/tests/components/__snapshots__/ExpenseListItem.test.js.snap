// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list item correctly 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/2"
  >
    <h3>
      Rent
    </h3>
  </Link>
  <p>
    $195.00
    -
    28th December, 1969
  </p>
</div>
`;
